File: app/src/main/java/org/dhis2/utils/jira/IssueRequest.java
package org.dhis2.utils.jira;

import org.dhis2.BuildConfig;

import java.util.ArrayList;

/**
 * QUADRAM. Created by ppajuelo on 11/04/2019.
 */
public class IssueRequest {

    private Fields fields;

    public IssueRequest(String summary, String description) {
        this.fields = new Fields(summary, description);
    }

    private class Fields {
        private Project project;
        private Issue issuetype;
        private String summary;
        private ArrayList<Component> components;
        private String description;
        private String environment;
        private ArrayList<FixVersion> affectedVersions;

        Fields(String summary, String description) {
            this.summary = summary;
            this.description = formatDescription(description);
            this.project = new Project("10200");
            this.issuetype = new Issue("Bug");
            this.components = new ArrayList<>();
            this.components.add(new Component("AndroidApp"));
            this.affectedVersions = new ArrayList<FixVersion>();
            this.affectedVersions.add(new FixVersion());
        }
    }

    private class Project {
        private String id;

        Project(String projectId) {
            this.id = projectId;
        }
    }

    private class Issue {
        private String name;

        Issue(String bug) {
            this.name = bug;
        }
    }

    private class Component {
        private String name;

        Component(String componentName) {
            name = componentName;
        }
    }

    private class FixVersion {
        String name = BuildConfig.VERSION_NAME;
    }

    public String formatDescription(String description) {
        return String.format("{panel:title=Bug description}\n%s\n{panel}", description);
    }
}

File: app/src/main/java/org/dhis2/utils/jira/IssueRequest.kt
package org.dhis2.utils.jira

import org.dhis2.BuildConfig

import java.util.ArrayList

/**
 * QUADRAM. Created by ppajuelo on 11/04/2019.
 */
class IssueRequest(summary: String, description: String) {

    private val fields: Fields

    init {
        this.fields = Fields(summary, description)
    }

    private inner class Fields internal constructor(private val summary: String, private val description: String) {
        private val project: Project = Project("10200")
        private val issuetype: Issue = Issue("Bug")
        private val components: ArrayList<Component> = ArrayList()
        private val fixVersions: ArrayList<FixVersion>

        init {
            this.components.add(Component("AndroidApp"))
            this.fixVersions = ArrayList()
            this.fixVersions.add(FixVersion())
        }


    }

    private inner class Project internal constructor(private val id: String)

    private inner class Issue internal constructor(private val name: String)

    private inner class Component internal constructor(private val name: String)

    private inner class FixVersion {
        internal var name = BuildConfig.VERSION_NAME
    }
}